website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,693 @@
+  unhandled.
+
+In Rust, `async` functions are just syntax sugar for a data structure that
+implements `Future`. Spawning a `Future` into a runtime[^1] like `tokio`
+requires this data structure to be `'static` and therefore, it cannot contain
+any references, including `&mut`. To mutate shared state in such a scenario, you
+basically have two options:
+
+- Use reference-counted pointers and a mutex, i.e. `Arc<Mutex<T>>`
+- Use "actors" and connect them via channels, i.e. spawn multiple tasks with
+  loops that read and write to channels
+
+Both of these options have a runtime overhead: Locks can result in contention
+and sending messages through channels requires copying. In addition, multiple
+tasks running inside a runtime operate in a non-deterministic order which can
+easily lead to race conditions and in the worst-case, dead-locks. It appears
+that with either of these options, we arrive at a design that feels brittle, is
+prone to dead-locks and no longer employs zero-cost abstractions, yet avoiding
+all of these is one of the reasons we wanted to use Rust in the first-place!
+
+In the sans-IO world, these problems don't exist. Our protocol code doesn't
+spawn any tasks and thus, `&mut self` is all we need to mutate state. Without
+tasks or threads, we also don't need synchronisation primitives like `Mutex`.
+Without channels, there is no need to copy data: The state machine can simply
+directly reference the buffer we passed to the socket.
+
+Last but not least, we've also found that ever since we moved to sans-IO, our
+code became much easier to understand. No more tracking down of: Where is the
+other end of this channel? What if the channel is closed? Which other code is
+locking this `Mutex`? Instead, it is all just nested state machines and regular
+function calls.
+
+## The downsides
+
+There are no silver-bullets and sans-IO is no exception to this. Whilst writing
+your own eventloop gives you great control, it can also result in subtle bugs
+that are initially hard to find.
+
+For example, a bug in the state machine where the value returned from
+`poll_timeout` is not advanced can lead to a busy-looping behaviour in the
+eventloop.
+
+Also, sequential workflows require more code to be written. In Rust, `async`
+functions compile down to state machines, with each `.await` point representing
+a transition to a different state. This makes it easy for developers to write
+sequential code together with non-blocking IO. Without `async`, we need to write
+our own state machines for expressing the various steps. How annoying this will
+be in practise depends on your problem domain. Modelling a request-response
+protocol is not very difficult as we've seen in the example of a `StunBinding`.
+On the other hand, if need to express larger, sequential workflows, manually
+modelling them out as state machines could become tedious.
+
+Finally, the sans-IO design is not particularly wide-spread (yet) in the Rust
```

Comment:
You could add a link to withoutboats – the maker of async Rust (or at least the `Pin` trait) blog for anyone who wants a deeper analysis. This one briefly discusses Sans-IO and str0m https://without.boats/blog/let-futures-be-futures/

I think boats' "intra task concurrency" advantages of async vs your thoughts on running multiple STUN requests is an interesting illustration of how Sans-IO can be an alternative to async.

Reply:
Added a short paragraph linking to the post!